VariantShredding.md

```diff
@@ -25,290 +25,316 @@
-For example, if `typed_value` is of type `int64`, then the string “123” might be contained in `variant_value`, which would not be reflected in statistics, but could be retained by a filter like `where variant_get(col, “$.field”, “long”) = 123`.
-If `variant_value` is all-null, then the engine can prune pages or row groups based on `typed_value`.
-This specification is not strict about what values may be stored in `variant_value` rather than `typed_value`, so it is not safe to skip rows based on `typed_value` unless the corresponding `variant_value` column is all-null, or the engine has specific knowledge of the behavior of the writer that produced the shredded data.
+Fields of an object can be shredded using a Parquet group for `typed_value` that contains shredded fields.
 
-# Shredding Semantics
+If the value is an object, `typed_value` must be non-null.
+If the value is not an object, `typed_value` must be null.
 
-Reconstruction of Variant value from a shredded representation is not expected to produce a bit-for-bit identical binary to the original unshredded value.
-For example, in a reconstructed Variant value, the order of object field values may be different from the original binary.
-This is allowed since the [Variant Binary Encoding](VariantEncoding.md#object-field-id-order-and-uniqueness) does not require an ordering of the field values, but the field IDs will still be ordered lexicographically according to the corresponding field names.
+Each shredded field in the `typed_value` group is represented as a required group that contains optional `value` and `typed_value` fields.
+The `value` field stores the value as Variant-encoded `binary` when the `typed_value` cannot represent the field.
+This layout enables readers to skip data based on the field statistics for `value` and `typed_value`.
 
-The physical representation of scalar values may also be different in the reconstructed Variant binary.
-In particular, the [Variant Binary Encoding](VariantEncoding.md) considers all integer and decimal representations to represent a single logical type.
-This flexibility enables shredding to be applicable in more scenarios, while maintaining all information and values losslessly.
-As a result, it is valid to shred a decimal into a decimal column with a different scale, or to shred an integer as a decimal, as long as no numeric precision is lost.
-For example, it would be valid to write the value 123 to a Decimal(9, 2) column, but the value 1.234 would need to be written to the `variant_value` column.
-When reconstructing, it would be valid for a reader to reconstruct 123 as an integer, or as a Decimal(9, 2).
-Engines should not depend on the physical type of a Variant value, only the logical type.
+If the value is a partially shredded object, the `value` must not contain the shredded fields.
+If shredded fields are present in the variant object, it is invalid and readers must either fail or use the shredded values.
```

Comment:
I think we should probably prefer failing.

Reply:
```suggestion
Readers are not required to check for overlap of between the fields in a `value` object and the `typed_value` schema.  If there is an overlap behavior of the reader is undefined.
```

This makes sense, I think we should just state the requirement that the check is not necessary and that it is undefined behavior if there is overlap.

Reply:
I like the part about not required to check, but the behavior cannot be undefined. If you don't have to check, then it should be a valid read and we need to know what value should be returned.